ops/main.py

```diff
@@ -172,10 +172,11 @@ class _JujuEvent:
     def __init__(self, charm_dir: Path):
         self._charm_dir = charm_dir
         self._exec_path = Path(sys.argv[0])
-        if self._exec_path.is_symlink() and self._exec_path.name != 'dispatch':
-            self._init_legacy()
-        else:
+
+        if 'JUJU_DISPATCH_PATH' in os.environ:
```

Comment:
So if you use this logic and only the symlink (or script) `hooks/install` exists (no 'dispatch'). In 2.8, You'll set the right path because juju told it to you, you'll then have self.is_dispatch set to True, so you'll see `hooks/install` as something that should be called. You'll either see it as a symlink, and won't call it, or you'll call it, but will have set OPERATOR_DISPATCH, so it won't do anything.
In 2.7, it won't be set, which means you'll init legacy, which is correct if it is a script or if it is a symlink. And you won't try to reinvoke.

And if 'dispatch' and 'hooks/install' exists. In 2.7 it won't touch dispatch, so the above logic applies. In 2.8 dispatch will trigger, see hooks/install, doesn't reinvoke for symlinks, reinvokes for scripts but will still have JUJU_DISPATCH_PATH set, and thus it will catch OPERATOR_DISPATCH.